tests/functional/src/main/java/org/mule/functional/classloading/isolation/classloader/MuleClassLoaderRunnerFactory.java

```diff
@@ -0,0 +1,265 @@
+     * @return the first Module from the list due to there should be only one module.
+     */
+    private MuleModule validatePluginModule(String pluginName, List<MuleModule> discoveredModules)
+    {
+        if (discoveredModules.size() == 0)
+        {
+            throw new IllegalStateException(pluginName + " doesn't have in its classpath a mule-module.properties to define what packages and resources should expose");
+        }
+        if (discoveredModules.size() > 1)
+        {
+            throw new IllegalStateException(pluginName + " has more than one mule-module.properties, composing plugins is not allowed");
+        }
+        return discoveredModules.get(0);
+    }
+
+    /**
+     * Creates an {@link ArtifactClassLoader} for the application.
+     *
+     * @param parent the parent class loader to be assigned to the new one created here
+     * @param childClassLoaderLookupPolicy look policy to be used
+     * @param artifactUrlClassification the url classifications to get plugins urls
+     * @return the {@link ArtifactClassLoader} to be used for test runner as in TCCL.
+     */
+    private ArtifactClassLoader createApplicationArtifactClassLoader(ClassLoader parent, ClassLoaderLookupPolicy childClassLoaderLookupPolicy, ArtifactUrlClassification artifactUrlClassification)
+    {
+        logClassLoaderUrls("APP", artifactUrlClassification.getApplicationURLs());
+        return new MuleArtifactClassLoader("app", artifactUrlClassification.getApplicationURLs().toArray(new URL[0]), parent, childClassLoaderLookupPolicy);
+    }
+
+    private void logClassLoaderUrls(final String classLoaderName, final List<URL> urls)
+    {
+        StringBuilder builder = new StringBuilder(classLoaderName).append(" classloader urls: [");
+        urls.stream().forEach(e -> builder.append("\n").append(" ").append(e.getFile()));
+        builder.append("\n]");
+        logClassLoadingTrace(builder.toString());
+    }
+
+    private void logClassLoadingTrace(String message)
+    {
+        if (isVerboseClassLoading())
+        {
+            logger.info(message);
+        }
+        else if (logger.isTraceEnabled())
+        {
+            logger.trace(message);
+        }
+    }
+
+    private Boolean isVerboseClassLoading()
+    {
+        return valueOf(getProperty(MULE_LOG_VERBOSE_CLASSLOADING));
+    }
+
+    private Set<String> getExtraBootPackages(Class<?> klass)
+    {
+        Set<String> packages = Sets.newHashSet();
+
+        List<String> extraBootPackagesList = getAnnotationAttributeFromHierarchy(klass, ArtifactClassLoaderRunnerConfig.class, "extraBootPackages");
+        extraBootPackagesList.stream().filter(extraBootPackages -> !isEmpty(extraBootPackages)).forEach(extraBootPackages -> addAll(packages, extraBootPackages.split(",")));
+
+        // Add default boot package always, they are defined in excluded.properties file!
+        try
+        {
+            Properties excludedProperties = RunnerModuleUtils.getExcludedProperties();
```

Comment:
Can you refactor the code so the excludedProperties is read in only one place? (Is also read on MuleClassPathClassifier#getExclusionsPredicate)